engine/src/logger/instance.cpp

```diff
@@ -12,277 +12,323 @@
 	#include <Windows.h>
 #endif
 
-namespace gen::logger {
-namespace {
-namespace fs = std::filesystem;
-
-void append_timestamp(std::string& out, Clock::time_point const& timestamp, Timestamp const mode) {
-	static auto s_mutex{std::mutex{}};
-	static constexpr std::size_t buf_size_v{64};
-	auto buffer = std::array<char, buf_size_v>{};
-	auto const time = Clock::to_time_t(timestamp);
-	// std::localtime / std::gmtime are not thread safe. cppref:
-	//  pointer to a static internal std::tm object on success, or null pointer otherwise.
-	// The structure may be shared between std::gmtime, std::localtime, and std::ctime,
-	// and may be overwritten on each invocation.
-	auto lock = std::unique_lock{s_mutex};
-	// NOLINTNEXTLINE
-	auto const tm_struct = mode == Timestamp::eUtc ? *std::gmtime(&time) : *std::localtime(&time);
-	lock.unlock();
-	std::strftime(buffer.data(), buffer.size(), "%F %T", &tm_struct);
-	out.append(buffer.data());
-}
-
-void append_location(std::string& out, std::source_location const& location, Location const mode) {
-	auto const path = [&] {
-		auto ret = fs::path{location.file_name()};
-		if (mode == Location::eFilename) { ret = ret.filename(); }
-		return ret.generic_string();
-	}();
-	out.append(path);
-}
-
-struct Formatter {
-	static constexpr auto open_v{'{'};
-	static constexpr auto close_v{'}'};
-
-	// snapshot of formatting-related Config (at the time of Formatter construction).
-	// we make copies instead of referring to Config to avoid mutex locks,
-	// since users can change the active config at any time.
-	struct Data {
-		// default format string size is 57, which will cause std::string to heap allocate.
-		// thus we use a fixed capacity stack string instead.
-		FixedString<Config::format_size_v> format{};
-		Location location{};
-		Timestamp timestamp{};
-	};
-
-	// NOLINTNEXTLINE
-	Data const& data;
-	// message and context don't need to be copied, they are unique per call stack.
-	std::string_view message;
-	// NOLINTNEXTLINE
-	Context const& context;
-
-	// output string (formatted)
-	std::string out{};
-	// remaining input
-	std::string_view format{data.format};
-	// current character (preceding format)
-	char current{};
-
-	[[nodiscard]] bool at_end() const { return format.empty(); }
-
-	bool advance() {
-		if (at_end()) {
-			// reset current
-			current = {};
-			return false;
+namespace gen::logger
+{
+	namespace
+	{
+		namespace fs = std::filesystem;
+
+		void append_timestamp(std::string & out, Clock::time_point const & timestamp, Timestamp const mode)
+		{
+			static auto s_mutex{std::mutex{}};
```

Comment:
Could you explain how you are avoiding the issues that come with std::gmtime here? Will the mutex be enough to ensure we don't have issues with the threads?

Reply:
`std::gmtime` and `std::localtime` may return a pointer to a shared an internal `static tm` instance, which is what makes it unsafe to use across multiple threads. We copy that returned `tm const*` by dereferencing it under a mutex lock, thus avoiding any data races.